dvc/external_repo.py

```diff
@@ -123,21 +123,35 @@ def _set_cache_dir(self):
 
         self.cache.local.cache_dir = cache_dir
 
-    def _set_upstream(self):
-        # check if the URL is local and no default remote is present
-        # add default remote pointing to the original repo's cache location
+    def _fix_upstream(self):
         if os.path.isdir(self.url):
-            if not self.config["core"].get("remote"):
-                src_repo = Repo(self.url)
-                try:
-                    cache_dir = src_repo.cache.local.cache_dir
-                finally:
-                    src_repo.close()
-
-                self.config["remote"]["auto-generated-upstream"] = {
-                    "url": cache_dir
-                }
-                self.config["core"]["remote"] = "auto-generated-upstream"
+            upstream_name = self.config["core"].get("remote")
+            src_repo = Repo(self.url)
+            try:
+                if upstream_name:
+                    self._fix_local_remote(src_repo, upstream_name)
+                else:
+                    self._add_upstream(src_repo)
+            finally:
+                src_repo.close()
+
+    def _fix_local_remote(self, src_repo, remote_name):
+        # Keep the relative path for a local remote relative
+        # to the original repo.
+        new_remote = self.config["remote"][remote_name]
+        old_remote = src_repo.config["remote"][remote_name]
+        # NOTE: it is assumed that only local+relative urls
+        # will be different from their old version. Revise this
+        # test if that stops being true.
```

Comment:
I would remove that. There already is an explanation of what this method is doing at the beginning. 
Comment `Revise...` will not make anyone revise that, that is the purpose of test :)